### PR TITLE
MQTT settings and telemetry

### PR DESCRIPTION
**Pull request by ryan-summers**
This PR replaces https://github.com/quartiq/stabilizer/pull/242

This PR addresses #149 by exposing a simple MQTT-based settings interface

**TODO**:
- [x] Resolve mutable ownership issues in the `MqttInterface`
- [x] Expand settings interface to `lockin.rs`
- [x] Add all settings (not yet supported by `miniconf`)
- [x] ~~Finalize and publish the `miniconf` package~~ Deferred to the future.
- [x] Move `MqttInterface`-like behaviors over to `miniconf`
- [x] Move smoltcp-nal to a separate crate
- [x] Test on hardware
- [x] ~~Implement stabilizer telemetry~~: Deferred to future

### CODE REVIEW

**Comment by jordens** [comment]
If there is an ergonomic advantage over using raw mqtt CLI utilities, let's also do a minimal CLI example client implementation for the IIR parameters in `dual-iir`.

File: Cargo.toml
Comment (ryan-summers) [inline Cargo.toml:56]: Update git dependencies to the quartiq repositories - also, publish `smoltcp-nal`

File: dsp/Cargo.toml
Comment (ryan-summers) [inline dsp/Cargo.toml:10]: I don't think we should need this. Might be a bug in miniconf that's requiring this

File: src/bin/dual-iir.rs
Comment (ryan-summers) [inline src/bin/dual-iir.rs:149]: This sleep seems to be behaving oddly - the network interface doesn't seem like it's waking up at the correct points and this needs some investigation.

File: src/bin/dual-iir.rs
Comment (ryan-summers) [inline src/bin/dual-iir.rs:177]: @jordens You had mentioned that we don't want to update the IIR channels all the time - would we always want to update the `iir_state` and `iir_channel` simultaneously? Also, would we want to update CH0/1 independently?

File: src/bin/lockin-external.rs
Comment (ryan-summers) [inline src/bin/lockin-external.rs:28]: Replace with settings similar to `dual-iir`

File: src/hardware/cycle_counter.rs
Comment (ryan-summers) [inline src/hardware/cycle_counter.rs:1]: Comments throughout this file

File: src/bin/dual-iir.rs
Comment (ryan-summers) [reply src/bin/dual-iir.rs:149]: Hmm, scratch that, appears to be working fine now

File: src/bin/dual-iir.rs
Comment (jordens) [reply src/bin/dual-iir.rs:177]: Yes. Only update the `iir_state` if that is explicitly flagged.

Also, this is somewhat related to https://github.com/quartiq/stabilizer/issues/70
Let me know if that description isn't clear.

File: dsp/src/iir.rs
Comment (jordens) [inline dsp/src/iir.rs:15]: The reason I wanted a newtype is to be able to implement things on it, like in `irr_int`.

Comment (ryan-summers) [reply dsp/src/iir.rs:15]: This type isn't supported by serde-json-core and it was why we were getting panics previously. We can't have anonymous fields in structs, or we will panic when deserializing them. If we want a custom type, we have to implement it with named fields.

That being said, perhaps we can redefine `Vec5` to
```rust
pub struct Vec5 {
    data: [f32; 5],
}
```

to get the best of both worlds?

Comment (jordens) [reply dsp/src/iir.rs:15]: Yes. That's fine!

Comment (jordens) [reply dsp/src/iir.rs:15]: In fact, maybe a trait for the type alias is better.

File: dsp/Cargo.toml
Comment (jordens) [inline dsp/Cargo.toml:12]: I think this is better done with a `patch` at the root. `dsp` shouldn't need `miniconf`, right?

Comment (ryan-summers) [reply dsp/Cargo.toml:12]: dsp needs `miniconf::StringSet` so that it can use it in the `#[derive(StringSet)]` for the DSP types, like IIR. If we don't want those to be directly settable, we can remove the dependency

Comment (jordens) [reply dsp/Cargo.toml:12]: Ah ok.
I was wondering when it would make sense to bring miniconf down into dsp. And this is just that case. Otherwise, we'd have to duplicate the IIR struct (structurally) in the settings.

Comment (jordens) [reply dsp/Cargo.toml:12]: But I still think the this should be done with a `patch` at the root Cargo.toml, no?

File: dsp/src/iir.rs
Comment (jordens) [reply dsp/src/iir.rs:15]: #274

File: dsp/Cargo.toml
Comment (ryan-summers) [reply dsp/Cargo.toml:12]: `patch` is generally used to override an existing dependency with a different version. In this case, `miniconf` isn't available at all in our dependency tree, so I'm not sure what we could `patch`. Unless you're referring to something else?

I dislike having to doubly specify the `miniconf` dependency, but I think that's a result of how we've structed `dsp` as a separate crate instead of a module (like `hardware`) within the `stabilizer` crate.

Comment (ryan-summers) [reply dsp/Cargo.toml:12]: Unless you're specifying the specify `branch` specifier to only appear at the root and use patch to push it down our dependency graph?

I plan on waiting until that branch is merged to mainline and we release miniconf before merging this, so I've been leaving it as-is for now.

Comment (jordens) [reply dsp/Cargo.toml:12]:
```
[dependencies]
miniconf = "0.1"
```

in both Cargo.toml and then

```
[patch.crates-io]
miniconf = { git = "https://github.com/quartiq/miniconf.git", branch = "rs/issue-21/terminal-array-elements"}
```

only at the root works afaict.

File: src/bin/lockin-external.rs
Comment (jordens) [inline src/bin/lockin-external.rs:38]: We don't necessarily need to bother with the other binaries in this PR yet. Having just `dual-iir` for now would be ok.

Comment (ryan-summers) [reply src/bin/lockin-external.rs:38]: I was thinking the same thing - otherwise, things start to get a big big. Let me yank all of the lockin-external changes to a different branch so we can keep things tidy.

File: src/bin/dual-iir.rs
Comment (jordens) [reply src/bin/dual-iir.rs:177]: The more I think about it, for now it's ok to never update `iir_state`. Let's keep it out of settings.

File: dsp/Cargo.toml
Comment (jordens) [reply dsp/Cargo.toml]: Yeah. It's a nit. Just a few fewer lines.

**Review by jordens — APPROVED** [review]

**Comment by ryan-summers** [comment]
I have bench tested setting various values of `dual-iir`, such as the IIR y_offset parameter and the AFE gain. Everything appears to be in working order for merging.

**Comment by ryan-summers** [comment]
bors merge

**Pull request merged**
